src/Hamiltonians/correlation_functions.jl

```diff
@@ -39,13 +43,28 @@ end
 
 LOStructure(::Type{<:G2RealCorrelator}) = IsDiagonal()
 
-function diagonal_element(::G2RealCorrelator{D}, add::SingleComponentFockAddress{N,M}) where {D,N,M}
+function diagonal_element(::G2RealCorrelator{D}, add::SingleComponentFockAddress) where {D}
+    M = num_modes(add)
+    d = mod(D, M)
+    v = onr(add)
+    if d == 0
+        return dot(v, v .- 1) / M
+    else
```

Comment:
This should be fine (I think) because the value of `d` can be computed from only type information. So the compiler should be able to remove the `if` condition in the compiled code. Maybe it is worth checking that this is true with `@code_llvm` or `@code_native`.

Reply:
I also think it is fine. But it is a good excuse to split it up properly